daemon/mgr/container.go

```diff
@@ -1215,118 +1217,197 @@ func (mgr *ContainerManager) updateContainerResources(c *Container, resources ty
 		}
-		return nil
 	}
-	// If container is running,  we need change
-	// configuration and recreate it. Else we just store new meta
-	// into disk, next time when starts container, the new configurations
-	// will take effect.
-
-	// Inherit volume configurations from old container,
-	// New volume configurations may cover the old one.
-	// c.VolumesFrom = []string{c.ID}
 
-	// FIXME(ziren): here will forcely stop container afer 3s.
-	// If DestroyContainer failed, we think the old container
-	// not changed, so just return error, no need recover it.
-	if _, err := mgr.Client.DestroyContainer(ctx, c.ID, 3); err != nil {
-		return errors.Wrapf(err, "failed to destroy container")
+	// prepare new snapshot for the new container
+	oldSnapID := c.SnapshotID()
+	newSnapID, err := mgr.prepareSnapshotForUpgrade(ctx, c.Key(), c.SnapshotID(), config.Image)
+	if err != nil {
+		return err
 	}
+	c.SetSnapshotID(newSnapID)
 
-	// remove snapshot of old container
-	if err := mgr.Client.RemoveSnapshot(ctx, c.ID); err != nil {
-		return errors.Wrap(err, "failed to remove snapshot")
+	// initialize container storage config before container started
+	if err := mgr.initContainerStorage(ctx, c); err != nil {
+		return errors.Wrapf(err, "failed to init container storage, id: (%s)", c.Key())
 	}
 
-	// wait util old snapshot to be deleted
-	wait := make(chan struct{})
-	go func() {
-		for {
-			// FIXME(ziren) Ensure the removed snapshot be removed
-			// by garbage collection.
-			time.Sleep(100 * time.Millisecond)
-
-			_, err := mgr.Client.GetSnapshot(ctx, c.ID)
-			if err != nil && errdefs.IsNotFound(err) {
-				close(wait)
-				return
+	// If container is running, we also should start the container
+	// after recreate it.
+	if c.IsRunning() {
+		if err := mgr.createContainerdContainer(ctx, c, "", ""); err != nil {
+			needRollback = true
+			if err := mgr.Client.RemoveSnapshot(ctx, newSnapID); err != nil {
+				logrus.Errorf("failed to remove snapshot %s: %v", newSnapID, err)
 			}
-		}
-	}()
-
-	select {
-	case <-wait:
-		// TODO delete snapshot succeeded
-	case <-time.After(30 * time.Second):
-		needRollback = true
-		return fmt.Errorf("failed to deleted old snapshot: wait old snapshot %s to be deleted timeout(30s)", c.ID)
-	}
 
-	// create a snapshot with image for new container.
-	if err := mgr.Client.CreateSnapshot(ctx, c.ID, config.Image); err != nil {
-		needRollback = true
-		return errors.Wrap(err, "failed to create snapshot")
+			return errors.Wrap(err, "failed to create new container")
+		}
 	}
 
-	if err := mgr.createContainerdContainer(ctx, c, "", ""); err != nil {
-		needRollback = true
-		return errors.Wrap(err, "failed to create new container")
+	// Upgrade success, remove snapshot of old container
+	if err := mgr.Client.RemoveSnapshot(ctx, oldSnapID); err != nil {
+		// TODO(ziren): remove old snapshot failed, may cause dirty data
+		logrus.Errorf("failed to remove snapshot %s: %v", oldSnapID, err)
```

Comment:
Do we need the log level of `Error`? Is `Warn` enough? 
Just double check to see the right way to use error and warn.